src/components/RecipeForm.test.tsx

```diff
@@ -0,0 +1,110 @@
+    const recipe = {
+      id: 1,
+      name: 'Pizza',
+      description: 'Put in the oven',
+      ingredients: [{
+        name: 'cheese'
+      },
+        {
+          name: 'mushrooms'
+        }]
+    }
+
+    it('should load successfully the form with the recipe values', async () => {
+      const onSubmitMock = jest.fn();
+      const onCancelMock = jest.fn();
+      render(<RecipeForm recipe={recipe} onSubmit={onSubmitMock} onCancel={onCancelMock}/>);
+
+      expect(screen.getByRole('button', {name: 'Save'})).toBeInTheDocument();
+      expect((screen.getByLabelText('Name') as HTMLInputElement).value).toBe('Pizza');
+      expect((screen.getByLabelText('Description') as HTMLInputElement).value).toBe('Put in the oven');
+      expect((screen.getByTestId('ingredient0') as HTMLInputElement).value).toBe('cheese');
+      expect((screen.getByTestId('ingredient1') as HTMLInputElement).value).toBe('mushrooms');
+    });
+
+    it('should submit the correct data when the user has removed an ingredient and clicks Save', async () => {
+      const onSubmitMock = jest.fn(e => e.preventDefault());
+      const onCancelMock = jest.fn();
+      render(<RecipeForm recipe={recipe} onSubmit={onSubmitMock} onCancel={onCancelMock}/>);
+
+      await act(async () => {
+        const firstRemoveButton = screen.getAllByRole('button', {name: "Remove"})[0]
+        fireEvent.click(firstRemoveButton);
+      });
+
+      await act(async () => {
+        fireEvent.click(screen.getByRole('button', {name: "Save"}));
+      });
+
+      expect(onSubmitMock).toHaveBeenCalledWith(expect.anything(), {...recipe, ingredients: [{name: 'mushrooms'}]})
+    });
+
+    it('should submit the correct data when the user has added a new ingredient and clicks Save', async () => {
+      const onSubmitMock = jest.fn(e => e.preventDefault());
+      const onCancelMock = jest.fn();
+      render(<RecipeForm recipe={recipe} onSubmit={onSubmitMock} onCancel={onCancelMock}/>);
+
+      await act(async () => {
+        fireEvent.click(screen.getByRole('button', {name: "Add Ingredient"}));
+      });
+
+      await act(async () => {
+        await userEvent.type(screen.getByTestId('ingredient2'), 'olives', {delay: 0.01});
+      });
+
+      await act(async () => {
+        fireEvent.click(screen.getByRole('button', {name: "Save"}));
+      });
+
+      expect(onSubmitMock).toHaveBeenCalledWith(expect.anything(), {
+        ...recipe,
+        ingredients: [{name: 'cheese'}, {name: 'mushrooms'}, {name: 'olives'}]
+      })
+    });
+
+    it('should submit the correct data when the user has modified a field and clicks Save', async () => {
+      const onSubmitMock = jest.fn(e => e.preventDefault());
+      const onCancelMock = jest.fn();
+      render(<RecipeForm recipe={recipe} onSubmit={onSubmitMock} onCancel={onCancelMock}/>);
+
+      await act(async () => {
+        await userEvent.type(screen.getByLabelText('Name'), ' Carbonara', {delay: 0.01});
```

Comment:
just curious, why did you decide add a delay there?

Reply:
Without adding the delay instead of getting the value Pizza Carbonara I was getting Pizzaa. So it was like the user had typed only the last character of the string passed